irohad/ametsuchi/impl/postgres_command_executor.cpp

```diff
@@ -703,84 +701,70 @@ namespace iroha {
-                                      (SELECT amount FROM src_amount LIMIT 1)
-                                      THEN
-                                      (SELECT amount FROM src_amount LIMIT 1)
-                                  ELSE 0::decimal
-                              END) - $6::decimal AS value
-                          ),
-              new_dest_value AS (SELECT
-                              (SELECT $6::decimal +
-                                  CASE WHEN EXISTS
-                                      (SELECT amount FROM dest_amount LIMIT 1)
-                                          THEN
-                                      (SELECT amount FROM dest_amount LIMIT 1)
-                                  ELSE 0::decimal
-                              END) AS value
-                          ),
+              new_src_quantity AS (
+                  SELECT coalesce(sum(amount), 0) - $6::decimal as value
+                  FROM account_has_asset
+                     WHERE asset_id = $4 AND
+                     account_id = $2
+              ),
+              new_dest_quantity AS (
+                  SELECT coalesce(sum(amount), 0) + $6::decimal as value
+                  FROM account_has_asset
+                     WHERE asset_id = $4 AND
+                     account_id = $3
+              ),
+              checks AS (
+                  -- error code and check result
+                  SELECT 3 code, count(1) = 1 result
+                  FROM account
+                  WHERE account_id = $2
+
+                  UNION
+                  SELECT 4, count(1) = 1
+                  FROM account
+                  WHERE account_id = $3
+
+                  UNION
+                  SELECT 5, count(1) = 1
+                  FROM asset
+                  WHERE asset_id = $4
+                     AND precision >= $5
+
+                  UNION
+                  SELECT 6, value >= 0
+                  FROM new_src_quantity
+
+                  UNION
+                  SELECT 7, value < (2::decimal ^ 256) / 10::decimal ^ precision
+                  FROM new_dest_quantity, asset
+                  WHERE asset_id = $4
+              ),
               insert_src AS
               (
-                  INSERT INTO account_has_asset(account_id, asset_id, amount)
-                  (
-                      SELECT $2, $4, value
-                      FROM new_src_value
-                      WHERE EXISTS (SELECT * FROM has_src_account LIMIT 1) AND
-                        EXISTS (SELECT * FROM has_dest_account LIMIT 1) AND
-                        EXISTS (SELECT * FROM has_asset LIMIT 1) AND
-                        EXISTS (SELECT value FROM new_src_value
-                                WHERE value >= 0 LIMIT 1) %s
-                  )
-                  ON CONFLICT (account_id, asset_id)
-                  DO UPDATE SET amount = EXCLUDED.amount
-                  RETURNING (1)
+                  UPDATE account_has_asset
```

Comment:
Please check if there is a test case when the receiver does not have the corresponding asset, because this query should not work in that case.

Reply:
it seems alright